Modules/selectmodule.c

```diff
@@ -298,30 +330,55 @@ select_select_impl(PyObject *module, PyObject *rlist, PyObject *wlist,
         tvp = &tv;
     }
 
-#ifdef SELECT_USES_HEAP
+#if defined(MS_WINDOWS) || defined(SELECT_USES_HEAP)
     /* Allocate memory for the lists */
-    rfd2obj = PyMem_NEW(pylist, FD_SETSIZE + 1);
-    wfd2obj = PyMem_NEW(pylist, FD_SETSIZE + 1);
-    efd2obj = PyMem_NEW(pylist, FD_SETSIZE + 1);
+    rfd2obj = PyMem_NEW(pylist, setsize + 1);
+    wfd2obj = PyMem_NEW(pylist, setsize + 1);
+    efd2obj = PyMem_NEW(pylist, setsize + 1);
     if (rfd2obj == NULL || wfd2obj == NULL || efd2obj == NULL) {
         if (rfd2obj) PyMem_Free(rfd2obj);
         if (wfd2obj) PyMem_Free(wfd2obj);
         if (efd2obj) PyMem_Free(efd2obj);
         return PyErr_NoMemory();
     }
-#endif /* SELECT_USES_HEAP */
+#else /* !defined(MS_WINDOWS) && !defined(SELECT_USES_HEAP) */
+    rfd2obj = (pylist*)&rfd2obj_stack;
+    wfd2obj = (pylist*)&wfd2obj_stack;
+    efd2obj = (pylist*)&efd2obj_stack;
+#endif /* defined(MS_WINDOWS) || defined(SELECT_USES_HEAP) */
+
+#if defined(MS_WINDOWS)
+    /* Allocate memory for the sets */
+    /* We ought to allocate setsize * sizeof(SOCKET) +
+    sizeof(int) bytes. With this approach on 64-bit we
+    allocate 4 additional unused bytes but the allocation
+    code looks a lot simpler.*/
+    ifdset = (fd_set*) PyMem_NEW(SOCKET, setsize + 1);
+    ofdset = (fd_set*) PyMem_NEW(SOCKET, setsize + 1);
+    efdset = (fd_set*) PyMem_NEW(SOCKET, setsize + 1);
+    if (ifdset->fd_array == NULL || ofdset->fd_array == NULL || efdset->fd_array == NULL) {
+        if (ifdset->fd_array) PyMem_DEL(ifdset->fd_array);
```

Comment:
PyMem_DEL() is a alias to PyMem_Free(): use it directly.

The NULL test is useless and can be removed. PyMem_Free(NULL) is well defined, it does nothing: https://docs.python.org/dev/c-api/memory.html#c.PyMem_Free